### PR TITLE
M4163 index transaction log

### PR DESCRIPTION
**Pull request by jjettenn**

### CODE REVIEW

File: molgenis-data/src/main/java/org/molgenis/DatabaseConfig.java [hunk 1]
```diff
@@ -65,9 +66,15 @@ public DataSource dataSource()
 	}
 
 	@Bean
+	public IdGenerator idGenerator()
```
Comment (bartcharbon) [inline molgenis-data/src/main/java/org/molgenis/DatabaseConfig.java]: why is this needed? there is an autowired idGenerator available in this class right?

Comment (jjettenn) [inline molgenis-data/src/main/java/org/molgenis/DatabaseConfig.java]: I got a NullPointerException on IdGenerator.

File: molgenis-data/src/main/java/org/molgenis/data/transaction/index/IndexTransactionLogRepositoryDecorator.java [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+	public void drop()
+	{
+		decorated.drop();
+	}
+
+	@Override
+	public void rebuildIndex()
+	{
+		indexTransactionLogService.log(getEntityMetaData(), CudType.UPDATE, DataType.METADATA, null);
+		decorated.rebuildIndex();
```
Comment (tommydeboer) [inline molgenis-data/src/main/java/org/molgenis/data/transaction/index/IndexTransactionLogRepositoryDecorator.java]: Why is the rebuilding of the index also transactional?

File: molgenis-data/src/main/java/org/molgenis/data/transaction/index/IndexTransactionLogMetaData.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+		addAttribute(INDEX_STATUS).setDataType(new EnumField()).setEnumOptions(IndexStatus.getOptions())
+				.setNillable(false);
+		addAttribute(END_TIME).setDataType(MolgenisFieldTypes.DATETIME).setNillable(true);
+		addAttribute(LOG_COUNT).setDataType(MolgenisFieldTypes.INT).setNillable(false);
+	}
+
+	/**
+	 * Transaction status
+	 * 
+	 * @author jonathanjetten
```
Comment (tommydeboer) [inline molgenis-data/src/main/java/org/molgenis/data/transaction/index/IndexTransactionLogMetaData.java]: Remove author tag (or put it in top of class)

File: molgenis-data/src/main/java/org/molgenis/data/transaction/index/IndexTransactionLogMetaData.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+			}
+
+			return options;
+		}
+	};
+
+	/**
+	 * Indexing transaction status
+	 * 
+	 * @author jonathanjetten
```
Comment (tommydeboer) [inline molgenis-data/src/main/java/org/molgenis/data/transaction/index/IndexTransactionLogMetaData.java]: Remove author tag (or put it in top of class)

File: molgenis-data/src/test/java/org/molgenis/data/transaction/index/IndexTransactionLogServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+		TransactionSynchronizationManager.unbindResource(MolgenisTransactionManager.TRANSACTION_ID_RESOURCE_NAME);
+	}
+
+	@Test
+	public void createLog()
+	{
+		Entity entity = indexTransactionLogService.createLog("1");
+		assertEquals(entity.get(IndexTransactionLogMetaData.TRANSACTION_ID), "1");
+		// assertEquals(entity.get(IndexTransactionLogMetaData.USER_NAME), null); // on server null on local machine
+		// "username"
```
Comment (tommydeboer) [inline molgenis-data/src/test/java/org/molgenis/data/transaction/index/IndexTransactionLogServiceTest.java]: what to do with this?

File: molgenis-integration-tests/src/main/java/org/molgenis/integrationtest/data/AbstractDataApiTestConfig.java [hunk 6]
```diff
@@ -22,6 +22,12 @@
 import org.molgenis.data.support.DataServiceImpl;
 import org.molgenis.data.support.OwnedEntityMetaData;
 import org.molgenis.data.support.UuidGenerator;
+<<<<<<< HEAD
+import org.molgenis.data.transaction.TransactionConfig;
+import org.molgenis.data.transaction.TransactionLogService;
+import org.molgenis.data.transaction.index.IndexTransactionLogService;
+=======
+>>>>>>> 83de4092ec8c103bbf0acc2758e9b8d27f0d8064
```
Comment (tommydeboer) [inline molgenis-integration-tests/src/main/java/org/molgenis/integrationtest/data/AbstractDataApiTestConfig.java]: There are still some merge artefacts here

File: molgenis-core-ui/src/main/java/org/molgenis/ui/MolgenisRepositoryDecoratorFactory.java [hunk 7]
```diff
@@ -23,6 +25,7 @@
 public class MolgenisRepositoryDecoratorFactory implements RepositoryDecoratorFactory
 {
 	private final EntityManager entityManager;
+	private final IndexTransactionLogService indexTansactionLogService;
```
Comment (Mark-de-Haan) [inline molgenis-core-ui/src/main/java/org/molgenis/ui/MolgenisRepositoryDecoratorFactory.java]: typo: indexTansactionLogService - > indexTransactionLogService

File: molgenis-data/src/main/java/org/molgenis/data/support/DefaultEntity.java [hunk 8]
```diff
@@ -31,6 +31,7 @@
 	private static final long serialVersionUID = 1L;
 
 	private final Map<String, Object> values = new LinkedCaseInsensitiveMap<>();
+
```
Comment (Mark-de-Haan) [inline molgenis-data/src/main/java/org/molgenis/data/support/DefaultEntity.java]: Don't commit this change

File: molgenis-data/src/main/java/org/molgenis/data/transaction/index/IndexTransactionLogRepositoryDecorator.java (hunk 2 above)
Comment (jjettenn) [inline molgenis-data/src/main/java/org/molgenis/data/transaction/index/IndexTransactionLogRepositoryDecorator.java]: Logging the transaction CUD operations for reindexing purposes

File: molgenis-data-elasticsearch/src/main/java/org/molgenis/data/elasticsearch/reindex/ReindexActionRegisterService.java [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+					int actionOrder = increaseCountReindexActionJob(reindexActionJob);
+					Entity reindexAction = this.createReindexAction(reindexActionJob, entityMetaData.getName(),
+							cudType, dataType, entityId, actionOrder);
+					dataService.add(ReindexActionMetaData.ENTITY_NAME, reindexAction);
+				});
+			}
+			else
+			{
+				LOG.debug("Transaction id is unknown");
```
Comment (fdlk) [inline molgenis-data-elasticsearch/src/main/java/org/molgenis/data/elasticsearch/reindex/ReindexActionRegisterService.java]: make that a warn I think, shouldn't happen I think?

File: molgenis-data-elasticsearch/src/main/java/org/molgenis/data/elasticsearch/reindex/ReindexActionJobMetaData.java [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+public class ReindexActionJobMetaData extends DefaultEntityMetaData
+{
+	public static final String ENTITY_NAME = "reindex_action_job";
+	public static final String ID = "id";
+	public static final String COUNT = "count";
+
+	public ReindexActionJobMetaData(String backend)
+	{
+		super(ENTITY_NAME);
+		setBackend(backend);
```
Comment (fdlk) [inline molgenis-data-elasticsearch/src/main/java/org/molgenis/data/elasticsearch/reindex/ReindexActionJobMetaData.java]: can you add a description for the entity, perhaps even also the attributes?

File: molgenis-data-elasticsearch/src/main/java/org/molgenis/data/elasticsearch/reindex/ReindexActionMetaData.java [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+	public static final String ACTION_ORDER = "action_order";
+	public static final String ENTITY_FULL_NAME = "entity_full_name";
+	public static final String ENTITY_ID = "entity_id";
+	public static final String CUD_TYPE = "cud_type";
+	public static final String DATA_TYPE = "data_type";
+	public static final String REINDEX_STATUS = "reindex_status";
+
+	public ReindexActionMetaData(ReindexActionJobMetaData indexTransactionLogMetaData, String backend)
+	{
+		super(ENTITY_NAME);
```
Comment (fdlk) [inline molgenis-data-elasticsearch/src/main/java/org/molgenis/data/elasticsearch/reindex/ReindexActionMetaData.java]: Same here, can you add a description?

File: molgenis-data-elasticsearch/src/main/java/org/molgenis/data/elasticsearch/reindex/ReindexActionRegisterService.java [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+
+	/**
+	 * Log and create locks for an add/update/delete operation on a Repository
+	 * 
+	 * @param entityMetaData
+	 * @param cudType
+	 */
+	public synchronized void register(EntityMetaData entityMetaData, CudType cudType, DataType dataType, String entityId)
+	{
+		if (!ReindexActionRegisterService.EXCLUDED_ENTITIES.contains(entityMetaData.getName()))
```
Comment (fdlk) [inline molgenis-data-elasticsearch/src/main/java/org/molgenis/data/elasticsearch/reindex/ReindexActionRegisterService.java]: If their repository isn't indexed, does it make sense to keep track of the changes?
(Then again, will all repositories be indexed in 2.0?)

File: molgenis-data-elasticsearch/src/main/java/org/molgenis/data/elasticsearch/reindex/ReindexActionRegisterService.java [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+		this.reindexActionMetaData = reindexActionMetaData;
+	}
+
+	/**
+	 * Log and create locks for an add/update/delete operation on a Repository
+	 * 
+	 * @param entityMetaData
+	 * @param cudType
+	 */
+	public synchronized void register(EntityMetaData entityMetaData, CudType cudType, DataType dataType, String entityId)
```
Comment (fdlk) [inline molgenis-data-elasticsearch/src/main/java/org/molgenis/data/elasticsearch/reindex/ReindexActionRegisterService.java]: won't making this method synchronized make the entire dataservice essentially single-threaded?

File: molgenis-data-elasticsearch/src/main/java/org/molgenis/data/elasticsearch/reindex/ReindexActionRepositoryCollectionDecorator.java [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+
+import org.molgenis.data.AttributeMetaData;
+import org.molgenis.data.Entity;
+import org.molgenis.data.EntityMetaData;
+import org.molgenis.data.ManageableRepositoryCollection;
+import org.molgenis.data.Repository;
+import org.molgenis.data.elasticsearch.reindex.ReindexActionMetaData.CudType;
+import org.molgenis.data.elasticsearch.reindex.ReindexActionMetaData.DataType;
+
+public class ReindexActionRepositoryCollectionDecorator implements ManageableRepositoryCollection
```
Comment (fdlk) [inline molgenis-data-elasticsearch/src/main/java/org/molgenis/data/elasticsearch/reindex/ReindexActionRepositoryCollectionDecorator.java]: Here also, add a little bit of javadoc to explain what this class does

File: molgenis-data-elasticsearch/src/main/java/org/molgenis/data/elasticsearch/reindex/ReindexActionRegisterService.java (hunk 9 above)
Comment (jjettenn) [inline molgenis-data-elasticsearch/src/main/java/org/molgenis/data/elasticsearch/reindex/ReindexActionRegisterService.java]: Okay

File: molgenis-data-elasticsearch/src/main/java/org/molgenis/data/elasticsearch/reindex/ReindexActionRepositoryDecorator.java [hunk 15]
```diff
@@ -0,0 +1,244 @@
...
+import org.molgenis.data.EntityMetaData;
+import org.molgenis.data.Fetch;
+import org.molgenis.data.Query;
+import org.molgenis.data.QueryRule.Operator;
+import org.molgenis.data.Repository;
+import org.molgenis.data.RepositoryCapability;
+import org.molgenis.data.elasticsearch.reindex.ReindexActionMetaData.CudType;
+import org.molgenis.data.elasticsearch.reindex.ReindexActionMetaData.DataType;
+
+public class ReindexActionRepositoryDecorator implements Repository<Entity>
```
Comment (fdlk) [inline molgenis-data-elasticsearch/src/main/java/org/molgenis/data/elasticsearch/reindex/ReindexActionRepositoryDecorator.java]: And here too, little bit of javadoc

File: molgenis-data/src/test/java/org/molgenis/data/transaction/MolgenisTransactionManagerTest.java [hunk 16]
```diff
@@ -4,22 +4,15 @@
...
 {
-	private MolgenisTransactionManager molgenisTransactionManager;
-
-	@BeforeMethod
-	public void beforeMethod()
-	{
-		molgenisTransactionManager = new MolgenisTransactionManager(new UuidGenerator(), mock(DataSource.class));
-	}
+	private MolgenisTransactionManager molgenisTransactionManager = new MolgenisTransactionManager(new UuidGenerator(),
+			mock(DataSource.class));
```
Comment (fdlk) [inline molgenis-data/src/test/java/org/molgenis/data/transaction/MolgenisTransactionManagerTest.java]: (why) is this needed?

File: molgenis-data-elasticsearch/src/main/java/org/molgenis/data/elasticsearch/reindex/ReindexActionMetaData.java [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+import java.util.List;
+
+import org.molgenis.MolgenisFieldTypes;
+import org.molgenis.data.support.DefaultEntityMetaData;
+import org.molgenis.fieldtypes.EnumField;
+import org.molgenis.fieldtypes.TextField;
+
+public class ReindexActionMetaData extends DefaultEntityMetaData
+{
+	public static final String ENTITY_NAME = "reindex_action";
```
Comment (fdlk) [inline molgenis-data-elasticsearch/src/main/java/org/molgenis/data/elasticsearch/reindex/ReindexActionMetaData.java]: The _ name scheme is not quite the same as our usual ReindexAction with attributes called reindexActionGroup and actionOrder. Is this on purpose?
